cache/test_h2_cache_control.py

```diff
@@ -1014,3 +1014,442 @@ def test_ignore_max_stale_in_request(self):
+
+        if expect_stale:
+            # expect stale response
+            self.assertTrue(int(resp_age) >= 3)
+
+            stale_warn = client.last_response.headers.get("warning")
+            self.assertTrue(stale_warn == "110 - Response is stale")
+        else:
+            # expect not cached response
+            self.assertTrue(resp_age == -1)
+
+
+class TestH2CacheUseStaleTimeout(TestH2CacheUseStaleBase):
+    backends = [
+        {
+            "id": "deproxy",
+            "type": "deproxy_unreliable",
+            "port": "8000",
+            "response": "static",
+            "response_content": "",
+        }
+    ]
+
+    def test_use_stale_timeout(self):
+        """
+        Send first request, this request forwarded to upstream and processed with status
+        code 200, Tempesta cache it. Wait while response become stale. Send second
+        request, at this time upstream is broken, it just disconnects without any
+        responses. After few tries Tempesta responds to client with stale cached
+        response.
+
+        We expect stale response, because set "cache_use_stale 504", without
+        this directive we would expect 504 response from Tempesta, due to forwarding
+        timeout.
+        """
+        server = self.get_server("deproxy")
+        server.hang_on_req_num = 2
+        self.use_stale_base(302, "504", [], [], "200", True)
+
+    def test_timeout_no_stale(self):
+        """
+        Send first request, this request forwarded to upstream and processed with status
+        code 200, Tempesta cache it. Wait while response become stale. Send second
+        request, at this time upstream is broken, it just disconnects without any
+        responses. After few tries Tempesta responds to client with 504 status code.
+
+        Stale response is not expected.
+        """
+        server = self.get_server("deproxy")
+        server.hang_on_req_num = 2
+        self.use_stale_base(302, "500", [], [], "504", False)
+
+
+class TestH2CacheUseStale(TestH2CacheUseStaleBase):
+    backends = [
+        {
+            "id": "deproxy",
+            "type": "deproxy",
+            "port": "8000",
+            "response": "static",
+            "response_content": "",
+        }
+    ]
+
+    @marks.Parameterize.expand(
+        [
+            marks.Param(name="500", status=500, use_stale="500", r2_hdrs=[]),
+            marks.Param(name="400", status=400, use_stale="400", r2_hdrs=[]),
+            marks.Param(name="5xx_500", status=500, use_stale="5*", r2_hdrs=[]),
+            marks.Param(name="5xx_502", status=502, use_stale="5*", r2_hdrs=[]),
+            marks.Param(name="4xx_400", status=400, use_stale="4*", r2_hdrs=[]),
+            marks.Param(name="4xx_403", status=403, use_stale="4*", r2_hdrs=[]),
+            marks.Param(name="4xx&5xx", status=404, use_stale="4* 5*", r2_hdrs=[]),
+            marks.Param(
+                name="invalid_response", status=403, use_stale="5*", r2_hdrs=[("hd>r", "v")]
```

Comment:
You didn't get the test right, here we expect 5xx status code generated by Tempesta during parsing invalid response, not upstream. If you specify `use_stale="4*"` you can't distinguish why we got stale, it was due to error during parsing or because we had 403 from upstream. I've added comment to the test.